detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/LongParameterList.kt

```diff
@@ -5,36 +5,72 @@ import io.gitlab.arturbosch.detekt.api.Debt
 import io.gitlab.arturbosch.detekt.api.ThresholdedCodeSmell
 import io.gitlab.arturbosch.detekt.rules.isOverride
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtConstructor
+import org.jetbrains.kotlin.psi.KtFunction
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.KtParameterList
+import org.jetbrains.kotlin.psi.KtPrimaryConstructor
+import org.jetbrains.kotlin.psi.KtSecondaryConstructor
 
 /**
  * Reports functions which have more parameters than a certain threshold (default: 6).
  *
  * @configuration threshold - number of parameters required to trigger the rule (default: `6`)
+ * (deprecated: "Use `functionThreshold` and `constructorThreshold` instead")
+ * @configuration functionThreshold - number of function parameters required to trigger the rule (default: `6`)
+ * @configuration constructorThreshold - number of constructor parameters required to trigger the rule (default: `7`)
  * @configuration ignoreDefaultParameters - ignore parameters that have a default value (default: `false`)
+ * @configuration ignoreDataClasses - ignore long constructor parameters list for data classes (default: `true`)
  *
  * @active since v1.0.0
  */
 class LongParameterList(
-    config: Config = Config.empty,
-    threshold: Int = DEFAULT_THRESHOLD_PARAMETER_LENGTH
-) : ThresholdRule(config, threshold) {
+    config: Config = Config.empty
+) : Rule(config) {
 
     override val issue = Issue("LongParameterList",
             Severity.Maintainability,
-            "The more parameters a method has the more complex it is. Long parameter lists are often " +
+            "The more parameters a function has the more complex it is. Long parameter lists are often " +
                     "used to control complex algorithms and violate the Single Responsibility Principle. " +
-                    "Prefer methods with short parameter lists.",
+                    "Prefer functions with short parameter lists.",
             Debt.TWENTY_MINS)
 
+    private val functionThreshold: Int =
+        valueOrDefault(FUNCTION_THRESHOLD, valueOrDefault(THRESHOLD, DEFAULT_FUNCTION_THRESHOLD))
+
+    private val constructorThreshold: Int =
+        valueOrDefault(CONSTRUCTOR_THRESHOLD, valueOrDefault(THRESHOLD, DEFAULT_CONSTRUCTOR_THRESHOLD))
+
     private val ignoreDefaultParameters = valueOrDefault(IGNORE_DEFAULT_PARAMETERS, false)
 
+    private val ignoreDataClasses = valueOrDefault(IGNORE_DATA_CLASSES, true)
+
     override fun visitNamedFunction(function: KtNamedFunction) {
+        validateFunction(function, functionThreshold)
+    }
+
+    override fun visitPrimaryConstructor(constructor: KtPrimaryConstructor) {
+        validateConstructor(constructor)
+    }
+
+    override fun visitSecondaryConstructor(constructor: KtSecondaryConstructor) {
+        validateConstructor(constructor)
+    }
+
+    private fun <T : KtConstructor<T>> validateConstructor(constructor: KtConstructor<T>) {
```

Comment:
I'm sorry, I don't understand why this function needs generics.
Does the following function declaration also suffice?

```suggestion
    private fun validateConstructor(constructor: KtFunction) {
```

Reply:
I've just taken least common denominator of two types, which happens to have generic parameter. More generic KtFunction should work, will check.

Reply:
`.getContainingClassOrObject()` is a method of `KtConstructor`, so it's not available on `KtFunction`, but if star-projected type is okay, that's more than enough for this.